### PR TITLE
Add Dynamic Config Service to Core Service

### PR DESCRIPTION
**Pull request by huyaboo**
### Description

This PR adds a `DynamicConfigService` to core, enabling a subset of configs to be dynamically determined at request time. See #7111 for detailed design. By default, the feature is "disabled", meaning that when calling `client.getConfig(...)`, the `dynamicConfigService` will return the static config in `opensearch_dashboards.yml`. The service will still exist as part of core

### TODO
There are several TODO items outside the scope of this PR:
- [ ] Register routes to make the client accessible to the service (this PR will not affect configs since the client will default to `yml` configs
- [ ] Limit configs that are `exposeToBrowser`
- [ ] Validation logic when creating configs
- [ ] Determine logic for clearing the cache. Right now, the cache will not be deleted
- [ ] Add migration logic to deprecations, much like how saved objects are migrated

### Issues Resolved
Partially addresses #7111 

## Screenshot
The below recording shows `csp.rules` being added as a document in the config store index `.opensearch_dashboards_config` (alias for `.opensearch_dashboards_config_1`). The site will intentionally break when the `csp.rules` is set to `default-src 'none'`.

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/73027756/2aa60b48-ce3a-4fd7-9e46-cb5895c57a75

**Note:** in the recording, I had routes registered to add routes. This is PoC code and is not included in this PR. This is to help demonstrate changes using the OpenSearch client.

## Changelog
- feat: Add DynamicConfigService as a new Core service

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by zhyuanqi** [comment]
Hi Huy. Overall looks good to me.  Could we add more description for opensearchstoreclient and DummyConfigStoreClient. Also in http server, you registered registerDynamicConfigHandlers for csp. Could you add more description on this as well.

File: src/core/server/config/service/config_store_client/opensearch_config_store_client.ts [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+              add: {
+                index: getDynamicConfigIndexName(1),
+                alias: DYNAMIC_APP_CONFIG_ALIAS,
+              },
+            },
+          ],
+        },
+      });
+    }
+  }
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/core/server/config/service/config_store_client/opensearch_config_store_client.ts:71]: The alias is necessary as in the future, migration logic will be implemented for the config index. The migration logic will ensure that upon OSD version upgrade, all existing configs within the index are up to date. The alias makes it easy to determine what the latest config index is. Think of it as a `.kibana_N` index of sorts.

**Comment by huyaboo** [comment]
> Hi Huy. Overall looks good to me. Could we add more description for opensearchstoreclient and DummyConfigStoreClient. Also in http server, you registered registerDynamicConfigHandlers for csp. Could you add more description on this as well.

Added some more descriptions to the config store clients

**Review by zhyuanqi — APPROVED** [review]

**Review by BionIT — DISMISSED** [review]

**Review by zhyuanqi — APPROVED** [review]

**Review by BionIT — APPROVED** [review]

**Review by zhyuanqi — APPROVED** [review]

**Comment by SuZhou-Joe** [comment]
#7911

**Pull request merged**